### ISSUE TITLE
Cannot provide oauth2 with a proper config causing crash on Google App Engine

### ISSUE THREAD

**Issue by orian** [go-authboss/authboss#77]
tl;dr; https://godoc.org/golang.org/x/oauth2#Config.Exchange on Google App Engine needs to get a `context.Context` from `appengine.NewContext(request)`

Longer: I've tried to deploy an app written as stand alone to GAE, worked fine, but when doing a oauth2 with Fb i'm getting an error. It's reproducible on a local GAE playground. The reason is that a GAE provides own http client and http.DefaultClient is blocked.

For now I'm using a hack:
 - a tiny middleware before authboss sets context through gorilla.context
 - a modified version of authboss checks if Context is available

I don't know if it worth fixing, but writing about it to save others time (took me few hours to precisely find the place of error). Maybe  just mention in the documentation.

In general, I think it's worth to consider the Google approach to handlers:
https://godoc.org/golang.org/x/net/context

A solution might be to allow to specify an option (a global in oauth2 subpackage):

```
var ContextProvider func(req *http.Request) context.Context { 
    return oauth2.NoContext
}
```
It would be easier than refactoring all code to provide third argument.

I've tested, second approach is nicer.
notice, one must import `google.golang.org/appengine` instead standard `appengine`

**Comment by orian** [comment]
Ok, the above was only a part of problem, the client has to be also provided to oauth2 Callback to fetch profile.

**Comment by aarondl** [comment]
Meditating on this issue, will probably accept the code as is. Just want to be sure about it. It is bringing in another dependency though. And there may be better ways to support GAE.

**Comment by orian** [comment]
I'm open for other solutions. The one I've introduced adds x/net/context, although it's not a standard it's widely used among Google's internal/external code.

The Context and client are a request specific objects, they have to be injected on each request. As I didn't want to depend on appengine package, i've decided to go for factory like function and allowing a user to configure it.

I've tested the approach described in first comment, but it was ugly.

**Comment by aarondl** [comment]
This PR was merged.
